teerdays/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+		BadState,
+	}
+
+	/// a store for all active bonds. tokentime is updated lazily
+	#[pallet::storage]
+	#[pallet::getter(fn teerday_bonds)]
+	pub type TeerDayBonds<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, TeerDayBondOf<T>, OptionQuery>;
+
+	/// a store for all pending unlocks which are awaiting the unlock period to pass.
+	/// Withdrawal happens lazily and causes entry removal from this store
+	#[pallet::storage]
+	#[pallet::getter(fn pending_unlock)]
+	pub type PendingUnlock<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, (T::Moment, BalanceOf<T>), OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(< T as Config >::WeightInfo::bond())]
+		pub fn bond(
+			origin: OriginFor<T>,
+			#[pallet::compact] value: BalanceOf<T>,
+		) -> DispatchResult {
+			let signer = ensure_signed(origin)?;
+			ensure!(!TeerDayBonds::<T>::contains_key(&signer), Error::<T>::AlreadyBonded);
+			ensure!(value >= T::Currency::minimum_balance(), Error::<T>::InsufficientBond);
+
+			let free_balance = T::Currency::free_balance(&signer);
+			let value = value.min(free_balance);
+			Self::deposit_event(Event::<T>::Bonded { account: signer.clone(), amount: value });
+			T::Currency::set_lock(TEERDAYS_ID, &signer, value, WithdrawReasons::all());
+			let teerday_bond = TeerDayBondOf::<T> {
+				value,
+				last_updated: pallet_timestamp::Pallet::<T>::get(),
+				accumulated_tokentime: BalanceOf::<T>::zero(),
+			};
+			TeerDayBonds::<T>::insert(&signer, teerday_bond);
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(< T as Config >::WeightInfo::bond())]
+		pub fn bond_extra(
+			origin: OriginFor<T>,
+			#[pallet::compact] value: BalanceOf<T>,
+		) -> DispatchResult {
+			let signer = ensure_signed(origin)?;
+			ensure!(value >= T::Currency::minimum_balance(), Error::<T>::InsufficientBond);
+			let bond = Self::do_update_teerdays(&signer)?;
+			let free_balance = T::Currency::free_balance(&signer);
+			// free confusingly includes the already bonded amount, so we need to subtract it
+			let value = value.min(free_balance.saturating_sub(bond.value));
+			let new_bond_value = bond.value.saturating_add(value);
+			Self::deposit_event(Event::<T>::Bonded { account: signer.clone(), amount: value });
+			T::Currency::set_lock(TEERDAYS_ID, &signer, new_bond_value, WithdrawReasons::all());
+			let teerday_bond = TeerDayBondOf::<T> {
+				value: new_bond_value,
+				last_updated: bond.last_updated,
+				accumulated_tokentime: bond.accumulated_tokentime,
+			};
+			TeerDayBonds::<T>::insert(&signer, teerday_bond);
+			Ok(())
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(< T as Config >::WeightInfo::unbond())]
+		pub fn unbond(
+			origin: OriginFor<T>,
+			#[pallet::compact] value: BalanceOf<T>,
+		) -> DispatchResult {
+			let signer = ensure_signed(origin)?;
+			ensure!(Self::pending_unlock(&signer).is_none(), Error::<T>::PendingUnlock);
```

Comment:
This might be todo for the future. I think it is not very convenient that we can't unbond more if we are already unbonding, but I do agree that this introduces quite some more complexity.

Reply:
yeah. the first implementation shall be of minimal complexity - even if that comes with some UX downsides